src/ngx_http_lua_worker_thread.c

```diff
@@ -0,0 +1,410 @@
+        put_task_ctx(myctx->ctx);
+        return;
+    }
+    c = r->connection;
+
+    ngx_http_lua_ctx_t *ctx;
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+    if (ctx == NULL) {
+        put_task_ctx(myctx->ctx);
+        return;
+    }
+
+    lua_State *vm = myctx->ctx->vm;
+
+    if (myctx->rc != 0) {
+        str = lua_tolstring(vm, 1, &len);
+        lua_pushboolean(L, myctx->rc);
+        lua_pushlstring(L, str, len);
+        nresults = 2;
+
+    } else {
+        /* copying return values */
+        lua_pushboolean(L, 1);
+        nresults = lua_gettop(vm);
+        for (i = 1; i <= nresults; i++) {
+            lauxh_xcopy(vm, L, i, 1);
+        }
+        nresults += 1;
+    }
+
+    put_task_ctx(myctx->ctx);
+
+    /* resume the caller coroutine */
+
+    myctx->wait_co_ctx->cleanup = myctx->cleanup;
+    myctx->wait_co_ctx->data = myctx->cleanup_data;
+    ctx->cur_co_ctx = myctx->wait_co_ctx;
+    vm = ngx_http_lua_get_lua_vm(r, ctx);
+
+    rc = ngx_http_lua_run_thread(vm, r, ctx, nresults);
+
+    ngx_log_debug1(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                   "lua run thread returned %d", rc);
+
+    if (rc == NGX_AGAIN) {
+        ngx_http_lua_run_posted_threads(c, vm, r, ctx, c->requests);
+        return;
+    }
+
+    if (rc == NGX_DONE) {
+        ngx_http_lua_finalize_request(r, NGX_DONE);
+        ngx_http_lua_run_posted_threads(c, vm, r, ctx, c->requests);
+        return;
+    }
+
+    /* rc == NGX_ERROR || rc >= NGX_OK */
+
+    if (ctx->entered_content_phase) {
+        ngx_http_lua_finalize_request(r, rc);
+        return;
+    }
+}
+
+
+static void
+ngx_http_lua_worker_thread_cleanup(void *data)
+{
+    my_thread_ctx_t *ctx = data;
+    ngx_thread_mutex_lock(&ctx->ctx->mutex, ngx_cycle->log);
+    ctx->is_abort = 1;
+    ngx_thread_mutex_unlock(&ctx->ctx->mutex, ngx_cycle->log);
+
+    if (ctx->cleanup) {
+        (ctx->cleanup)(ctx->cleanup_data);
+    }
+}
+
+
+static int
+ngx_http_lua_run_worker_thread(lua_State *L)
+{
+    ngx_http_request_t      *r;
+    ngx_http_lua_ctx_t      *ctx;
+    int                      n_args;
+
+    ngx_str_t                thread_pool_name;
+    ngx_thread_pool_t       *thread_pool;
+
+    task_ctx_t              *tctx;
+    lua_State               *vm;
+
+    size_t                   len;
+    const char              *mod_name;
+    int                      rc;
+
+    size_t                   len2;
+    const char              *err;
+
+    int                      i;
+
+    ngx_thread_task_t       *task;
+    my_thread_ctx_t         *myctx;
+
+    r = ngx_http_lua_get_req(L);
+    if (r == NULL) {
+        return luaL_error(L, "no request found");
+    }
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+    if (ctx == NULL) {
+        return luaL_error(L, "no ctx found");
+    }
+
+    n_args = lua_gettop(L);
+    if (n_args < 2) {
+        return luaL_error(L,
+            "expecting at least 2 arguments, but got %d", n_args);

```

Comment:
I have some questions about test cases.
* How to run test cases locally?
* How to add directives in nginx.conf main block? Because this api needs to use `thread_pool`